### PR TITLE
Artifact Registry for Kyma modules

### PR DESCRIPTION
**Pull request by the1bit**
**Description**

Changes proposed in this pull request:

- [Provide private image registry for cap-operator](https://github.tools.sap/kyma/backlog/issues/4079)
Notes: CAP operator is going open source we can skip the part with private registry for cap-operator-manager. The second part (private registry for module templates) is still valid, as we have other teams that will keep module templates private.

**Related issue(s)**

### CODE REVIEW

File: configs/terraform/modules/artifact-registry/README.md [hunk 1]
```diff
@@ -11,14 +11,14 @@ These are the default values:
...
 artifact_registry_multi_region   = true
 artifact_registry_primary_area  = "europe"
 ```
 
-When you use the GCP private image registry, consider the following: 
+When you use the GCP private image registry, consider the following:
 
 - The solution is prepared for the GCP Service Account related execution.
-- The `roles/artifactregistry.writer` role binding is part of the solution. To learn more, read [Artifact Registry Repository Access Control](https://cloud.google.com/artifact-registry/docs/access-control).
+- The `roles/artifactregistry.writer` role binding is part of the solution. To learn more, read [Artifact Registry Repository Access Control](https://cloud.google.com/artifact-registry/docs/access-control). (If this variable is empty the solution won't add any service account with writer permission)
```
Comment (IwonaLanger) [inline configs/terraform/modules/artifact-registry/README.md:21]:
```suggestion
- The `roles/artifactregistry.writer` role binding is part of the solution. To learn more, read [Artifact Registry Repository Access Control](https://cloud.google.com/artifact-registry/docs/access-control). If this variable is empty, the solution won't add any service account with the writer permission.
```

File: configs/terraform/modules/artifact-registry/README.md [hunk 2]
```diff
@@ -11,14 +11,14 @@ These are the default values:
...
+- The `roles/artifactregistry.reader` role binding is required for lifecycle-manager service accounts. To learn more, read [Artifact Registry Repository Access Control](https://cloud.google.com/artifact-registry/docs/access-control).
+- You can make repository to public if you use the `artifact_registry_public = true` in the `terraform.tfvars`
```
Comment (IwonaLanger) [inline configs/terraform/modules/artifact-registry/README.md:23]:
```suggestion
- You can make your repository public if you use the `artifact_registry_public = true` in the `terraform.tfvars`
```

File: configs/terraform/environments/dev/artifact-registry/main.tf [hunk 3]
```diff
@@ -1,14 +1,13 @@
...
-  artifact_registry_module         = var.artifact_registry_module
-  immutable_artifact_registry      = var.immutable_artifact_registry
-  artifact_registry_multi_region   = var.artifact_registry_multi_region
-  artifact_registry_owner          = var.artifact_registry_owner
-  artifact_registry_serviceaccount = var.artifact_registry_serviceaccount
-
-
+  source = "../../../modules/artifact-registry"
 
+  for_each                                 = { for i, v in var.artifact_registry_list : i => v }
```
Comment (dekiel) [inline configs/terraform/environments/dev/artifact-registry/main.tf:4]: Can you use a toset here, instead iterating over list.

```suggestion
  for_each                                 = toset(var.artifact_registry_list)
```

File: configs/terraform/environments/dev/artifact-registry/terraform.tfvars [hunk 4]
```diff
@@ -1,6 +1,9 @@
...
\ No newline at end of file
+gcp_region = "europe-west4"
+artifact_registry_list = [
+  {
+    name                   = "modules-internal"
+    owner                  = "neighbors"
+    type                   = "development"
+    reader_serviceaccounts = ["klm-controller-manager@sap-ti-dx-kyma-mps-dev.iam.gserviceaccount.com", "klm-controller-manager@sap-ti-dx-kyma-mps-stage.iam.gserviceaccount.com", "klm-controller-manager@sap-ti-dx-kyma-mps-prod.iam.gserviceaccount.com"]
+  },
+]
```
Comment (dekiel) [inline configs/terraform/environments/dev/artifact-registry/terraform.tfvars:9]:
```suggestion
gcp_region = "europe-west4"
artifact_registries = {
  modules-internal = {
    name                   = "modules-internal"
    owner                  = "neighbors"
    type                   = "development"
    reader_serviceaccounts = ["klm-controller-manager@sap-ti-dx-kyma-mps-dev.iam.gserviceaccount.com", "klm-controller-manager@sap-ti-dx-kyma-mps-stage.iam.gserviceaccount.com", "klm-controller-manager@sap-ti-dx-kyma-mps-prod.iam.gserviceaccount.com"]
  },
}
```
Alternatively, can we use a map of objects and pass a raw variable to `for_each`

```
for_each = var.artifact_registries
```

File: configs/terraform/environments/dev/artifact-registry/variables.tf [hunk 5]
```diff
@@ -12,45 +12,18 @@ variable "gcp_project_id" {
...
+    owner                  = string
+    type                   = string
+    writer_serviceaccount  = optional(string, "")
+    reader_serviceaccounts = list(string)
+    primary_area           = optional(string, "europe")
+    multi_region           = optional(bool, true)
+    public                 = optional(bool, false)
+    immutable              = optional(bool, false)
+  }))
 }
```
Comment (dekiel) [inline configs/terraform/environments/dev/artifact-registry/variables.tf:29]: A variable declaration for map of objects.

```suggestion
variable "artifact_registries" {
  description = "Artifact Registry related data set"
  type = map(object({
    name                   = string
    owner                  = string
    type                   = string
    writer_serviceaccount  = optional(string, "")
    reader_serviceaccounts = list(string)
    primary_area           = optional(string, "europe")
    multi_region           = optional(bool, true)
    public                 = optional(bool, false)
    immutable              = optional(bool, false)
  }))
}

```

**Review by dekiel — CHANGES_REQUESTED** [review]
Take a look on my proposal of implementation with map of objects. Is it possible? With it, we can simply pass a variable to `for_each`. It also gives an easy access to any element of the map.

File: configs/terraform/modules/artifact-registry/README.md [hunk 6]
```diff
@@ -1,24 +1,61 @@
...
+| owner                  | Registry Owner Team                                                     | string       | x        |               |
+| type                   | Environment type (development, production)                              | string       | x        |               |
+| reader_serviceaccounts | List of Service Accounts who have `Reader` access on registry           | list(string) | x        |               |
+| writer_serviceaccount  | List of Service Account who has  `Writer`  access on registry           | string       |          | ""            |
+| primary_area           | Primary area (if multi region registry)                                 | string       |          | europe        |
+| multi_region           | Multi region or single region registry                                  | bool         |          | true          |
+| public                 | Is it available for every users from the internet with `Reader` access? | bool         |          | false         |
+| immutable              | Immutable tags?                                                         | bool         |          | false         |
```
Comment (IwonaLanger) [inline configs/terraform/modules/artifact-registry/README.md:50]:
```suggestion
| **name**                   | Artifact Registry name                                                  | string       | x        |               |
| **owner**                  | Registry Owner Team                                                     | string       | x        |               |
| **type**                   | Environment type (development, production)                              | string       | x        |               |
| **reader_serviceaccounts** | List of Service Accounts that have `Reader` access to registry           | list(string) | x        |               |
| **writer_serviceaccount**  | List of Service Accounts that have  `Writer`  access to registry           | string       |          | ""            |
| **primary_area**           | Primary area (if multi-region registry)                                 | string       |          | europe        |
| **multi_region**           | Multi-region or single-region registry                                  | bool         |          | true          |
| **public**                 | Is it available for every internet user with `Reader` access? | bool         |          | false         |
| **immutable**              | Immutable tags?                                                         | bool         |          | false         |
```

Comment (IwonaLanger) [inline configs/terraform/modules/artifact-registry/README.md:50]: What does the description mean? Is the meaning: "Are there (any) immutable tags?"

File: configs/terraform/modules/artifact-registry/README.md [hunk 7]
```diff
@@ -1,24 +1,61 @@
...
+
+
+
+
+When you use the GCP private image registry, consider the following:
```
Comment (IwonaLanger) [inline configs/terraform/modules/artifact-registry/README.md:55]:
```suggestion

When you use the GCP private image registry, consider the following:
```

File: configs/terraform/modules/artifact-registry/README.md [hunk 8]
```diff
@@ -1,24 +1,61 @@
...
 
 - The solution is prepared for the GCP Service Account related execution.
-- The `roles/artifactregistry.writer` role binding is part of the solution. To learn more, read [Artifact Registry Repository Access Control](https://cloud.google.com/artifact-registry/docs/access-control).
-- Vulnerability scanning is enabled by default.
-- You must define `artifact_registry_serviceaccount` in the .tfvars file.
+- The `roles/artifactregistry.writer` role binding is part of the solution. To learn more, read [Artifact Registry Repository Access Control](https://cloud.google.com/artifact-registry/docs/access-control). If this variable is empty, the solution won't add any service account with the `writer` permission.
```
Comment (IwonaLanger) [inline configs/terraform/modules/artifact-registry/README.md:58]:
```suggestion
- The `roles/artifactregistry.writer` role binding is part of the solution. To learn more, read [Artifact Registry Repository Access Control](https://cloud.google.com/artifact-registry/docs/access-control). If this variable is empty, the solution doesn't add any service account with the `writer` permission.
```

File: configs/terraform/modules/artifact-registry/README.md (hunk 6 above)
Comment (the1bit) [reply configs/terraform/modules/artifact-registry/README.md:50]: This is a special tag. Theoretically it is known by the guys who use it. :-)

**Review by dekiel — APPROVED** [review]

**Review by IwonaLanger — APPROVED** [review]

**Pull request merged**